src/femtomes.cpp

```diff
@@ -0,0 +1,420 @@
+			}
+			else
+			{
+				_decode_state = FemtoDecodeState::pream_ble1;
+			}
+			break;
+
+		case FemtoDecodeState::pream_ble3:
+			if (temp == FEMTO_PREAMBLE3)
+			{
+				_decode_state = FemtoDecodeState::head_length;
+			}
+			else
+			{
+				_decode_state = FemtoDecodeState::pream_ble1;
+			}
+			break;
+
+		case FemtoDecodeState::head_length:
+			_femto_msg.header.data[0] = FEMTO_PREAMBLE1;
+			_femto_msg.header.data[1] = FEMTO_PREAMBLE2;
+			_femto_msg.header.data[2] = FEMTO_PREAMBLE3;
+			_femto_msg.header.data[3] = temp;
+			_femto_msg.header.femto_header.headerlength = temp;
+			_decode_state = FemtoDecodeState::head_data;
+			_femto_msg.read = 4;
+			break;
+
+		case FemtoDecodeState::head_data:
+			if (_femto_msg.read >= sizeof(_femto_msg.header.data)) {
+				_decode_state = FemtoDecodeState::pream_ble1;
+				break;
+			}
+			_femto_msg.header.data[_femto_msg.read] = temp;
+			_femto_msg.read++;
+			if (_femto_msg.read >= _femto_msg.header.femto_header.headerlength)
+			{
+				_decode_state = FemtoDecodeState::data;
+			}
+			break;
+		case FemtoDecodeState::data:
+			if (_femto_msg.read >= FEMO_MSG_MAX_LENGTH) {
+				_decode_state = FemtoDecodeState::pream_ble1;
+				break;
+			}
+			_femto_msg.data[_femto_msg.read - _femto_msg.header.femto_header.headerlength] = temp;
+			_femto_msg.read++;
+			if (_femto_msg.read >= (_femto_msg.header.femto_header.messagelength + _femto_msg.header.femto_header.headerlength))
+			{
+				_decode_state = FemtoDecodeState::crc1;
+			}
+			break;
+		case FemtoDecodeState::crc1:
+			_femto_msg.crc = (uint32_t) (temp << 0);
+			_decode_state = FemtoDecodeState::crc2;
+			break;
+		case FemtoDecodeState::crc2:
+			_femto_msg.crc += (uint32_t) (temp << 8);
+			_decode_state = FemtoDecodeState::crc3;
+			break;
+		case FemtoDecodeState::crc3:
+			_femto_msg.crc += (uint32_t) (temp << 16);
+			_decode_state = FemtoDecodeState::crc4;
+			break;
+		case FemtoDecodeState::crc4:
+			_femto_msg.crc += (uint32_t) (temp << 24);
+			_decode_state = FemtoDecodeState::pream_ble1;
+
+			uint32_t crc = calculateBlockCRC32((uint32_t)_femto_msg.header.femto_header.headerlength, (uint8_t *)&_femto_msg.header.data, (uint32_t)0);
+			crc = calculateBlockCRC32((uint32_t)_femto_msg.header.femto_header.messagelength, (uint8_t *)&_femto_msg.data[0], crc);
+
+			if (_femto_msg.crc == crc)
+			{
+				iRet = _femto_msg.read;
+			}
+			else
+			{
+				FEMTO_DEBUG("Femto: data packet is bad");
+			}
+			break;
+	}
+	return iRet;
+}
+
+void GPSDriverFemto::decodeInit()
+{
+	_decode_state = FemtoDecodeState::pream_ble1;
+
+}
+
+int GPSDriverFemto::writeAckedCommandFemto(const char* command, const char* reply, const unsigned int timeout)
+{
+	/**< write command*/
+	write(command, strlen(command));
+	/**< wait for reply*/
+	uint8_t buf[GPS_READ_BUFFER_SIZE];
+	gps_abstime time_started = gps_absolute_time();
+	while (time_started + timeout * 1000 * 2 > gps_absolute_time()) {
+		int ret = read(buf, sizeof(buf), 1000); /**< wait 1000us */

```

Comment:
Why is there a factor of ` * 2` above?
The timeout is in ms, and should be used from the argument:
```suggestion
		int ret = read(buf, sizeof(buf), timeout);

```

Reply:
thanks,this is my fault,not need * 2

Reply:
@bkueng the log of faster movements need some time.